opt: add WITH ORDINALITY via a RowNumber operator

justinj opened the pull request:
WITH ORDINALITY is has the behaviour of a special case of the window
function RANK with the empty set of partitioning columns. This commit
leaves most of window functions unimplemented, only implementing enough
to do WITH ORDINALITY.

WITH ORDINALITY introduces a new column `ordinality` to the input set,
whose value corresponds to a given row's position in that input set.
The semantics of the ordering are briefly discussed at
https://www.cockroachlabs.com/docs/stable/query-order.html#order-preservation.

This commit introduces window functions in a very limited form - only as
required for WITH ORDINALITY, so no partitioning or window functions
besides ROW_NUMBER() are supported. In addition, a given window function
can only have a single windowing operator.

Release note: None

Commit:
opt: add WITH ORDINALITY via a RowNumber operator

WITH ORDINALITY introduces a new column `ordinality` to the input set,
whose value corresponds to a given row's position in that input set.
The semantics of the ordering are briefly discussed at
https://www.cockroachlabs.com/docs/stable/query-order.html#order-preservation.

The original version of this commit introduced WITH ORDINALITY as a
special case of a window function, but it became clear that giving
window functions the full attention they require wouldn't be in the
immediate future. In addition, a RowNumber operator will be useful for
various decorrelations.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25154)
<!-- Reviewable:end -->

knz commented:
Wait if you do it this way, then during physical planning you'd be required to store all the rows upfront in the window frame before starting to output results (at least that's our current implementation). We lose the ability to produce the ordinality column in a streaming fashion. What do you do about that?

knz commented:
Please disregard my last comment, I hadn't seen the collapsing of window functions to ordinalityNode in `(ee *execEngine) ConstructWindowBy(input exec.Node)`. That alleviates my concerns.

knz commented:
Not only does this alleviate my concern, this is actually a good optimization because it also benefits queries specified via a windowing aggregation with `row_number` -- these collapse to WITH ORDINALITY in the physical plan and become distributable, whereas previously anthing window-related would not be distributable.

rytaft commented:
Reviewed 19 of 19 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/sql/opt_exec_engine.go, line 381 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIWtAeBwXJ2NmPFWfN:-LBIWtAeBwXJ2NmPFWfO:byoyqzl) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt_exec_engine.go#L381)):*
> ```Go
> 	copy(cols, inputColumns)
> 	cols[len(cols)-1] = sqlbase.ResultColumn{
> 		Name: "ordinality",
> ```

I'd either:

1. Rename this function to call it `ConstructOrdinality`, or
2. Pass this extra column in as a parameter so `ConstructWindowBy` can later be generalized for other window functions

---

*[pkg/sql/opt/metadata_column.go, line 131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIWMRd25PjGDphQ57c:-LBIWMRd25PjGDphQ57d:b-p4m66o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/metadata_column.go#L131)):*
> ```Go
> func (wk *WeakKeys) Copy() WeakKeys {
> 	res := make(WeakKeys, len(*wk))
> 	copy(res, *wk)
> ```

Don't you also need to call `.Copy()` on each weak key? Or does this not need to be a deep copy?

---

*[pkg/sql/opt/exec/execbuilder/testdata/window-by, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIYVejBWpwFBVXRVy5:-LBIYVejBWpwFBVXRVy6:b-ry2l9f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/exec/execbuilder/testdata/window-by#L15)):*
> ```
> 
> exec
> SELECT * FROM foo WITH ORDINALITY
> ```

why is it not showing column `x`?

---

*[pkg/sql/opt/memo/private_storage.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBI_MsS13pcx6NpoaL2:-LBI_MsS13pcx6NpoaL3:b-ro3cvv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/memo/private_storage.go#L206)):*
> ```Go
> }
> 
> func (ps *privateStorage) internWindowByDef(def *WindowByDef) PrivateID {
> ```

add a comment similar to other intern functions.

---

*[pkg/sql/opt/memo/testdata/logprops/window-by, line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIa3HrAAtuTVZt5A4J:-LBIa3HrAAtuTVZt5A4K:b-yw2129) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/memo/testdata/logprops/window-by#L14)):*
> ```
> ----
> window-by
>  ├── columns: ordinality:3(int!null)
> ```

you should include all columns here, not just ordinality

---

*[pkg/sql/opt/optbuilder/select.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIaR2iCMhXrjvAi5D3:-LBIaR2iCMhXrjvAi5D4:beye9b3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/optbuilder/select.go#L144)):*
> ```Go
> }
> 
> ////////
> ```

did you mean to add a comment here?

---

*[pkg/sql/opt/xform/testdata/physprops/ordering, line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIbsvmAfl4TzS8ZZBi:-LBIbsvmAfl4TzS8ZZBj:b6p5ef) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/xform/testdata/physprops/ordering#L287)):*
> ```
>  │         ├── best: (project G2 G3)
>  │         └── cost: 1000.00
>  ├── G2: (window-by G4 G5 ordering=)
> ```

if there is no ordering maybe don't print "ordering="

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25154)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt_exec_engine.go, line 381 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIWtAeBwXJ2NmPFWfN:-LBMEHJi7cXVVpjmGY7c:b-zc95ll) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt_exec_engine.go#L381)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I'd either:

1. Rename this function to call it `ConstructOrdinality`, or
2. Pass this extra column in as a parameter so `ConstructWindowBy` can later be generalized for other window functions
</blockquote></details>

Yes, I think this should be `ConstructOrdinality`. We've tried to keep most of the "smartness" (in this case detecting whether we can use the more efficient `ordinalityNode`) in the `execbuilder`, not the factory.

---

*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBMEdWS9Sp2vxiY39bC:-LBMEdWTCDasQ8tPTgxb:bolrfkm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/exec/execbuilder/relational_builder.go#L486)):*
> ```Go
> 	def := ev.Private().(*memo.WindowByDef)
> 	if !def.PartitionCols.Empty() || ev.Child(1).Operator() != opt.RowNumberOp {
> 		panic("window functions besides WITH ORDINALITY are not implemented")
> ```

return error, don't panic

---

*[pkg/sql/opt/ops/scalar.opt, line 540 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBMGt8T7LGodk10B6ch:-LBMGt8T7LGodk10B6ci:b7jn2od) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/ops/scalar.opt#L540)):*
> ```
> 
> # RowNumber assigns an increasing integer, starting from 1, to successive rows
> # in a partition.
> ```

[nit] mention that it's in the context of windowing functions (otherwise it's not clear what "partition" refers to).

---

*[pkg/sql/opt/optbuilder/select.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBMHS0uCVUGAJK46DHb:-LBMHS0uCVUGAJK46DHc:bbr42l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/optbuilder/select.go#L146)):*
> ```Go
> ////////
> func (b *Builder) buildOrdinality(inScope *scope) (outScope *scope) {
> 	outScope = inScope.push()
> ```

I am not sure we want to `push` here. I think this is what's causing `*` to not select any of the input columns.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25154)*
<!-- Sent from Reviewable.io -->

justinj commented:
After these discussions I decided to put the windowing version of this on the backburner, and just implement a general `RowNumber` operator, which Andy said would also be useful for various decorrelations.

---

Reviewed 1 of 22 files at r2.
Review status: 1 of 17 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/sql/opt_exec_engine.go, line 381 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIWtAeBwXJ2NmPFWfN:-LBMbtQoCN8D-wWLATTQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt_exec_engine.go#L381)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yes, I think this should be `ConstructOrdinality`. We've tried to keep most of the "smartness" (in this case detecting whether we can use the more efficient `ordinalityNode`) in the `execbuilder`, not the factory.
</blockquote></details>

Done.

---

*[pkg/sql/opt/metadata_column.go, line 131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIWMRd25PjGDphQ57c:-LBMbxBo9LAqJeiPjiTp:b-szbpbq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/metadata_column.go#L131)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Don't you also need to call `.Copy()` on each weak key? Or does this not need to be a deep copy?
</blockquote></details>

The `WeakKeys` code read to me as if the individual weak keys were to be considered immutable, @RaduBerinde you wrote this code initially, right? Do you think of them that way?

---

*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBMEdWS9Sp2vxiY39bC:-LBMdIsV2W_ddZ336USy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/exec/execbuilder/relational_builder.go#L486)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

return error, don't panic
</blockquote></details>

Done.

---

*[pkg/sql/opt/memo/private_storage.go, line 206 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBI_MsS13pcx6NpoaL2:-LBMeHb91TVqBAKp1VDm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/memo/private_storage.go#L206)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

add a comment similar to other intern functions.
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/select.go, line 144 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIaR2iCMhXrjvAi5D3:-LBMeaG30ZeZ0b99_GvU:bpkenqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/optbuilder/select.go#L144)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

did you mean to add a comment here?
</blockquote></details>

Yes, thanks for catching!

---

*[pkg/sql/opt/optbuilder/select.go, line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBMHS0uCVUGAJK46DHb:-LBMqZtu2Hd-91iwhpnD:b-v7zyzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/optbuilder/select.go#L146)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I am not sure we want to `push` here. I think this is what's causing `*` to not select any of the input columns.
</blockquote></details>

Fixed, makes sense.

---

*[pkg/sql/opt/xform/testdata/physprops/ordering, line 287 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIbsvmAfl4TzS8ZZBi:-LBMsRM9ELMfhAV7aq6m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/xform/testdata/physprops/ordering#L287)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

if there is no ordering maybe don't print "ordering="
</blockquote></details>

Done.

---

*[pkg/sql/opt/exec/execbuilder/testdata/window-by, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIYVejBWpwFBVXRVy5:-LBN9h_z1HgvPfbAd_TB:bt2vr0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/exec/execbuilder/testdata/window-by#L15)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

why is it not showing column `x`?
</blockquote></details>

Fixed!

---

*[pkg/sql/opt/memo/testdata/logprops/window-by, line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBIa3HrAAtuTVZt5A4J:-LBMeFay1bk1rfUCxfAC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/memo/testdata/logprops/window-by#L14)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

you should include all columns here, not just ordinality
</blockquote></details>

Done.

---

*[pkg/sql/opt/ops/scalar.opt, line 540 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBMGt8T7LGodk10B6ch:-LBMeEj3Ctwm_FMzYX93:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e154961c5b4784e3f8bbed72d48f5f40591f4ce2/pkg/sql/opt/ops/scalar.opt#L540)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] mention that it's in the context of windowing functions (otherwise it's not clear what "partition" refers to).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-:-LBN5X1K0JE6GaHrpzH0:b-aa4lfi)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 19 of 22 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/optbuilder/select.go, line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBNEtym1P9JuMZyiQun:-LBNEtyn0EJXjQeG2vT8:b-7j6hd1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/optbuilder/select.go#L150)):*
> ```Go
> // See Builder.buildStmt for a description of the remaining input and
> // return values.
> func (b *Builder) buildRowNumber(inScope *scope, colName string) (outScope *scope) {
> ```

In the other buildXXX functions, `inScope` is the last param. It would be good to stay consistent with that convention.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-:-LBNFm1l39jUQFexDxku:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/memo/logical_props_builder.go, line 456 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBQw-fSC_fqBzJXtaxX:-LBQw-fSC_fqBzJXtaxY:b-1137rf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/memo/logical_props_builder.go#L456)):*
> ```Go
> 	props.Relational.WeakKeys = props.Relational.WeakKeys.Copy()
> 	props.Relational.WeakKeys.Add(util.MakeFastIntSet(int(def.ColID)))
> ```

This PR needs to be setting statistics for this new operator. Rebecca can give you guidance on how that's done.

In particular, how come `colStatFromChildren` is not blowing up, since this op is not in that switch statement? Maybe points to missing test(s)?

---

*[pkg/sql/opt/memo/private_defs.go, line 121 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBNBzNO4pgj8IysezBM:-LBQxw1G42Dvm2HirsOn:b-eql6co) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/memo/private_defs.go#L121)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Do we need the name in the def? We can look up the label from the metadata using the ColID..
</blockquote></details>

Agree, the name shouldn't be stored here, but instead fetched from Metadata in `execbuilder/relational_builder.go`. See `buildProject` for example of how that's done, since `Project` also needs to get column names.

---

*[pkg/sql/opt/memo/private_defs.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBQynuS-J_yPwLAig6V:-LBQynuS-J_yPwLAig6W:beiuzu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/memo/private_defs.go#L124)):*
> ```Go
> }
> 
> // CanProvideOrdering returns true if the windowing operator returns rows
> ```

NIT: windowing operator => row number operator

---

*[pkg/sql/opt/memo/private_defs.go, line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBR-xFH56nEeHkZMeAa:-LBR-xFH56nEeHkZMeAb:b-1627sp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/memo/private_defs.go#L132)):*
> ```Go
> 	// ascending by it, so any ordering columns after an ascending ordinality are
> 	// irrelevant.
> 	// TODO(justin): This could probably be generalized to some helper - when we
> ```

Can you add a TODO in `physicalPropsBuilder.canProvideOrdering` to do this? I think it'd be best to put the key checking code at the top of that method. It would trim any ordering columns from the ordering spec that came after columns that constitute a key for that op.

---

*[pkg/sql/opt/optbuilder/select.go, line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBR19tqDlgqTMw9H6P8:-LBR19tqDlgqTMw9H6P9:b-f5x5fp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/optbuilder/select.go#L150)):*
> ```Go
> // See Builder.buildStmt for a description of the remaining input and
> // return values.
> func (b *Builder) buildRowNumber(inScope *scope, colName string) (outScope *scope) {
> ```

NIT: I'd call this `buildWithOrdinality` because that's what SQL calls it. We remap to a different name in opt expressions, but the convention in the optbuilder is to use the names of the build "input" rather than the build "output".

---

*[pkg/sql/opt/optbuilder/testdata/ordinality, line 22 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBR2p5PCHeVaOmd0Zt-:-LBR2p5PCHeVaOmd0Zt0:b2rt7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/optbuilder/testdata/ordinality#L22)):*
> ```
>       └── c int
> 
> build
> ```

How about test cases that use `ORDER BY` + `WITH ORDINALITY`? I'd expect ordering by ordinality as well as ordering by other cols. In particular, you're focused on making sure the right columns are put in scope.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25154)*
<!-- Sent from Reviewable.io -->

justinj commented:
Review status: 2 of 19 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/exec/execbuilder/testdata/ordinality, line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBNBNaJ3U33Nj3l2M0v:-LBR6gcN1Nr03sBAGrZL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/exec/execbuilder/testdata/ordinality#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

will need updating, we now do `exec` and `EXPLAIN (VERBOSE) ..`
</blockquote></details>

Done.

---

*[pkg/sql/opt/memo/logical_props_builder.go, line 456 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBQw-fSC_fqBzJXtaxX:-LBRxCiD-Toz6DLD5Op-:b-xuqlmt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/memo/logical_props_builder.go#L456)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This PR needs to be setting statistics for this new operator. Rebecca can give you guidance on how that's done.

In particular, how come `colStatFromChildren` is not blowing up, since this op is not in that switch statement? Maybe points to missing test(s)?
</blockquote></details>

Done, PTAL for statistics stuff @rytaft!

---

*[pkg/sql/opt/memo/private_defs.go, line 121 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBNBzNO4pgj8IysezBM:-LBRDcw588PxSZC8fM6o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/memo/private_defs.go#L121)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Agree, the name shouldn't be stored here, but instead fetched from Metadata in `execbuilder/relational_builder.go`. See `buildProject` for example of how that's done, since `Project` also needs to get column names.
</blockquote></details>

Done.

---

*[pkg/sql/opt/memo/private_defs.go, line 124 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBQynuS-J_yPwLAig6V:-LBRDkFaEauU4CRy_t7Z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/memo/private_defs.go#L124)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: windowing operator => row number operator
</blockquote></details>

Done.

---

*[pkg/sql/opt/memo/private_defs.go, line 132 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBR-xFH56nEeHkZMeAa:-LBREOoTDwMzZHGHwW1C:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/memo/private_defs.go#L132)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Can you add a TODO in `physicalPropsBuilder.canProvideOrdering` to do this? I think it'd be best to put the key checking code at the top of that method. It would trim any ordering columns from the ordering spec that came after columns that constitute a key for that op.
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/select.go, line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBNEtym1P9JuMZyiQun:-LBREprJ4CkZknC4D-ul:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/optbuilder/select.go#L150)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

In the other buildXXX functions, `inScope` is the last param. It would be good to stay consistent with that convention.
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/select.go, line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBR19tqDlgqTMw9H6P8:-LBRF50NBn9YvQCoysv7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/optbuilder/select.go#L150)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: I'd call this `buildWithOrdinality` because that's what SQL calls it. We remap to a different name in opt expressions, but the convention in the optbuilder is to use the names of the build "input" rather than the build "output".
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/testdata/ordinality, line 22 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBR2p5PCHeVaOmd0Zt-:-LBRF6FNAP8db6Fues9o:bwyqjr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/optbuilder/testdata/ordinality#L22)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

How about test cases that use `ORDER BY` + `WITH ORDINALITY`? I'd expect ordering by ordinality as well as ordering by other cols. In particular, you're focused on making sure the right columns are put in scope.
</blockquote></details>

Done. I had some in the memo, but good to have some here too.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25154)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 18 of 19 files at r4.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/exec/execbuilder/testdata/ordinality, line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBNBNaJ3U33Nj3l2M0v:-LBRyY7FEwOpq_-htDE4:b-ihj7cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/exec/execbuilder/testdata/ordinality#L28)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

can use the `hide-colnames` directive after `exec` if you don't want to see the column names at the top

---

*[pkg/sql/opt/memo/statistics_builder.go, line 983 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBRz3SW9Xo8nFxOz683:-LBRz3SW9Xo8nFxOz684:bogkmx9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc4556f83bd5cd6087b4d5bb26454b37a707dd69/pkg/sql/opt/memo/statistics_builder.go#L983)):*
> ```Go
> 
> func (sb *statisticsBuilder) buildRowNumber(inputStats *opt.Statistics) {
> 	*sb.s = *inputStats
> ```

I think you should just set the row count -- I think we want to avoid reusing the column stats maps across operators

---

*[pkg/sql/opt/memo/statistics_builder.go, line 998 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBS-sIs5a8hJoN4XnEj:-LBS-sIs5a8hJoN4XnEk:bdrckxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc4556f83bd5cd6087b4d5bb26454b37a707dd69/pkg/sql/opt/memo/statistics_builder.go#L998)):*
> ```Go
> 	if colSet.Contains(int(def.ColID)) {
> 		// The ordinality column is a key, so every row is distinct.
> 		colStat.DistinctCount = uint64(sb.ev.Logical().Relational.Stats.RowCount)
> ```

I think the row count is already a uint64 - this will probably give you a linter error

---

*[pkg/sql/opt/memo/testdata/stats/ordinality, line 28 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBS0YbVBCChKdU4m6wu:-LBS0YbVBCChKdU4m6wv:b-z5b5m7) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc4556f83bd5cd6087b4d5bb26454b37a707dd69/pkg/sql/opt/memo/testdata/stats/ordinality#L28)):*
> ```
> 
> build
> SELECT * FROM (SELECT * FROM a WITH ORDINALITY) WHERE ordinality > 0 AND ordinality <= 10
> ```

This is good, but can you also add a similar test with a predicate like `WHERE y > 0 AND y <= 10` so that you test that you're correctly retrieving the stats from the input?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25154)*
<!-- Sent from Reviewable.io -->

justinj commented:
Sorry, thought I published these drafts yesterday but I guess they didn't go through!

---

Review status: 16 of 19 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/exec/execbuilder/testdata/ordinality, line 28 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBNBNaJ3U33Nj3l2M0v:-LBWBw87AGOrTTjDWhUt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3751418279112d12807933ea8ac11194c5b317b0/pkg/sql/opt/exec/execbuilder/testdata/ordinality#L28)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

can use the `hide-colnames` directive after `exec` if you don't want to see the column names at the top
</blockquote></details>

Done.

---

*[pkg/sql/opt/memo/statistics_builder.go, line 983 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBRz3SW9Xo8nFxOz683:-LBWCVVi6NN7HDAXIMdo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc4556f83bd5cd6087b4d5bb26454b37a707dd69/pkg/sql/opt/memo/statistics_builder.go#L983)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think you should just set the row count -- I think we want to avoid reusing the column stats maps across operators
</blockquote></details>

Done.

---

*[pkg/sql/opt/memo/statistics_builder.go, line 998 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBS-sIs5a8hJoN4XnEj:-LBWCYTeCxX3-1ymciwm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc4556f83bd5cd6087b4d5bb26454b37a707dd69/pkg/sql/opt/memo/statistics_builder.go#L998)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think the row count is already a uint64 - this will probably give you a linter error
</blockquote></details>

Done.

---

*[pkg/sql/opt/memo/testdata/stats/ordinality, line 28 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-LBS0YbVBCChKdU4m6wu:-LBWCljm5UoU5u2jnId0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc4556f83bd5cd6087b4d5bb26454b37a707dd69/pkg/sql/opt/memo/testdata/stats/ordinality#L28)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

This is good, but can you also add a similar test with a predicate like `WHERE y > 0 AND y <= 10` so that you test that you're correctly retrieving the stats from the input?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-:-LBaFwF5EH-j9tWTxysN:b2qzpna)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 16 of 19 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25154#-:-LBbZRoB18rmGtHunYBZ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

justinj commented:
TFTRs!

bors r+
